### ISSUE TITLE
Illegal storage access with oveloaded static[T] generic procs (works with 0.17.0, breaks with the devel branch)

### ISSUE THREAD

**Issue by jxy** [nim-lang/Nim#6174]
The following code works with the master branch or the release 0.17.0, but causes `SIGSEGV: Illegal storage access.` with the latest `devel` branch.

``` nim
type
  G[N:static[int],T] = object
    o: T

proc newG[N:static[int],T](x:var G[N,T], y:T) =
  x.o = y

# This works with 0.17.0 but not the latest devel branch.
# The following overloaded proc breaks the Nim compiler.
# The code compiles with the following commented out.
proc newG[T](n:static[int],x:T):auto = discard

var x:G[0,float]
x.newG 0.1
```

Back trace from `nim_temp`:
```
Traceback (most recent call last)
nim.nim(121)             nim
nim.nim(77)              handleCmdLine
main.nim(163)            mainCommand
main.nim(74)             commandCompileToC
modules.nim(240)         compileProject
modules.nim(180)         compileModule
passes.nim(217)          processModule
passes.nim(135)          processTopLevelStmt
sem.nim(550)             myProcess
sem.nim(522)             semStmtAndGenerateGenerics
semstmts.nim(1839)       semStmt
semexprs.nim(818)        semExprNoType
semexprs.nim(2258)       semExpr
semexprs.nim(725)        semIndirectOp
semexprs.nim(2244)       semExpr
semexprs.nim(801)        semDirectOp
semexprs.nim(656)        semOverloadedCallAnalyseEffects
semcall.nim(388)         semOverloadedCall
semcall.nim(211)         resolveOverloads
semcall.nim(91)          pickBestCandidate
sigmatch.nim(2232)       matches
sigmatch.nim(2189)       matchesAux
sigmatch.nim(1937)       paramTypesMatch
sigmatch.nim(326)        paramTypesMatchAux
sigmatch.nim(987)        typeRelImpl
sigmatch.nim(1589)       typeRelImpl
sigmatch.nim(928)        typeRelImpl
system.nim(3600)         failedAssertImpl
system.nim(3592)         raiseAssert
system.nim(2712)         sysFatal
Error: unhandled exception: 
  not (aOrig == nil)  [AssertionError]
```

of version:
```
$ nim_temp --version
Nim Compiler Version 0.17.1 (2017-08-02) [MacOSX: amd64]
Copyright (c) 2006-2017 by Andreas Rumpf

git hash: 28737e9a403655293d2d7ece716892df99e356a6
active boot switches:
```

**Comment by TiberiumN** [comment]
I don't think that second example should work, because it doesn't return anything at all.
But yeah, it's still a compiler crash, which is bad

**Comment by jxy** [comment]
@TiberiumN This is not about calling and running the second proc.  You can define that proc whatever you like.  The point is that the overloading resolution breaks the compiler.

**Comment by jxy** [comment]
git bisect gives me the following
```
367d23235182ad9468d1e83e9380d2eea0c134b1 is the first bad commit                                      
commit 367d23235182ad9468d1e83e9380d2eea0c134b1    
Author: Zahary Karadjov <zahary@gmail.com>         
Date:   Sat Jun 10 23:14:25 2017 +0300             

    fix #1017; fix #3309 

```

**Comment by jxy** [comment]
I also took advice from @TiberiumN and changed the test to
``` nim
type                     
  G[N:static[int],T] = object                     
    o: T                 

proc newG[N:static[int],T](x:var G[N,T], y:T) =   
  x.o = y                

# This works with 0.17.0 but not the latest devel branch.                                            
# The following overloaded proc breaks the Nim compiler.                                             
# The code compiles with the following commented out.                                                
proc newG[T](n:static[int],x:T):auto = G[n,T](o:x)                                                   

var x:G[0,float]         
x.newG 0.1
```

**Comment by zah** [comment]
Thanks for narrowing this down, @jxy, I'll probably be able to look at the issue soon.

**Comment by Yardanico** [comment]
FYI: Doesn't crash the compiler anymore

**Comment by jxy** [comment]
Looks like it is fixed before the release of 0.18.0.  I'm closing this issue.
